setup ratelimit e2e tests

zirain opened the pull request:
init e2e test base on gateway-api conformance test suite, related to https://github.com/kubernetes-sigs/gateway-api/pull/1899

Fixes: https://github.com/envoyproxy/gateway/issues/863

cc @arkodg

zirain commented on go.mod:
```diff
@@ -2,6 +2,8 @@ module github.com/envoyproxy/gateway
 
 go 1.20
 
+replace sigs.k8s.io/gateway-api => github.com/zirain-dev/gateway-api v0.0.0-20230401020600-647bb66b0479
```
removed after https://github.com/kubernetes-sigs/gateway-api/pull/1899 merged.

zirain commented on go.mod:
```diff
@@ -2,6 +2,8 @@ module github.com/envoyproxy/gateway
 
 go 1.20
 
+replace sigs.k8s.io/gateway-api => github.com/zirain-dev/gateway-api v0.0.0-20230401074556-93436f059be6
```
gateway-api v0.6.2 with https://github.com/kubernetes-sigs/gateway-api/pull/1899

zirain commented on test/e2e/tests/ratelimit.go:
```diff
@@ -0,0 +1,51 @@
+			gwNN := types.NamespacedName{Name: "same-namespace", Namespace: ns}
+			gwAddr := kubernetes.GatewayAndHTTPRoutesMustBeAccepted(t, suite.Client, suite.TimeoutConfig, suite.ControllerName, kubernetes.NewGatewayRef(gwNN), routeNN)
+
+			// TODO: find a better to make sure ratelimit load new configuration
```
cc @arkodg do you have any good idea.

arkodg replied:
as discussed on slack, one way to do this would be to check the values programmed in the RL Service using the HTTP endpoint `/rlconfig` https://github.com/envoyproxy/ratelimit#debug-port
For now I think this is fine and we can optimize in a follow up PR

Is it possible to reduce the sleep to 5s or 10s ?

arkodg replied:
okay, lets raise a GH issue to track this and ensure there is no slowness in config translation and distribution, maybe CP Observability in v0.5.0 can help with this

zirain replied:
xref: https://github.com/envoyproxy/gateway/issues/1253

arkodg commented on examples/redis/envoy-gateway-config.yaml:
```diff
@@ -0,0 +1,18 @@
+apiVersion: v1
```
currently we seem to be combining multiple resources into a single file
e.g. https://github.com/envoyproxy/gateway/blob/main/examples/kubernetes/grpc-routing.yaml

lets do the same here and call the file `examples/kubernetes/global-ratelimit.yaml` ?

arkodg commented on test/e2e/e2e_test.go:
```diff
@@ -0,0 +1,55 @@
+	require.NoError(t, v1beta1.AddToScheme(client.Scheme()))
+	require.NoError(t, egv1a1.AddToScheme(client.Scheme()))
+
+	t.Logf("Running conformance tests with %s GatewayClass\n cleanup: %t\n debug: %t\n enable all features: %t \n supported features: [%v]\n exempt features: [%v]",
```
```suggestion
	t.Logf("Running e2e tests with %s GatewayClass\n cleanup: %t\n debug: %t\n enable all features: %t \n supported features: [%v]\n exempt features: [%v]",
```

arkodg commented on test/e2e/e2e_test.go:
```diff
@@ -0,0 +1,55 @@
+	})
+
+	cSuite.Setup(t)
+	t.Logf("Running %d conformance tests", len(tests.ConformanceTests))
```
lets be consistent on naming - either `E2E` or `ImplementationSpecificConformance` or .... ?

arkodg commented on test/e2e/tests/ratelimit.go:
```diff
@@ -0,0 +1,51 @@
+			// just wait a bit more time for now
+			time.Sleep(60 * time.Second)
+
+			http.MakeRequestAndExpectEventuallyConsistentResponse(t, suite.RoundTripper, suite.TimeoutConfig, gwAddr, http.ExpectedResponse{
```
imho we should send exactly 3 requests, and then expect a 429

arkodg commented:
thanks @zirain, this PR is a massive addition for the project to improve e2e and reduce regressions for our Extension APIs

arkodg commented:
thanks for addressing comments @zirain , this PR adds a fork for the gateway-api repo which im hoping we dont need, and the upstream PR you raised is merged and can be brought in to this project soon

zirain commented:
> thanks for addressing comments @zirain , this PR adds a fork for the gateway-api repo which im hoping we dont need, and the upstream PR you raised is merged and can be brought in to this project soon

yes, we should not include my fork repo. let's wait upstream PR merged.

arkodg commented:
another option is to write our own utils/suite package and we can write a custom `Setup` with a custom kubectl apply but we leverage the existing `Run` 
https://github.com/kubernetes-sigs/gateway-api/blob/main/conformance/utils/suite/suite.go

zirain commented:
> another option is to write our own utils/suite package and we can write a custom `Setup` with a custom kubectl apply but we leverage the existing `Run` https://github.com/kubernetes-sigs/gateway-api/blob/main/conformance/utils/suite/suite.go

unless https://github.com/kubernetes-sigs/gateway-api/pull/1899 is unacceptable.

zirain commented:
cc @arkodg , this's ready to go, PTAL.

Xunzhuo commented on tools/make/kube.mk:
```diff
@@ -96,6 +96,25 @@ kube-demo-undeploy: ## Uninstall the Kubernetes resources installed from the `ma
+	kubectl rollout restart deployment envoy-gateway -n envoy-gateway-system
+	kubectl wait --timeout=5m -n envoy-gateway-system deployment/envoy-gateway --for=condition=Available
+
+.PHONY: run-e2e
```
How about `run-e2e-ratelimit` ? Since we might need to add other e2e tests like auth, etc.

zirain replied:
this's the entrypoint for all e2e tests, for now there's only ratelimit tests

arkodg commented on tools/make/kube.mk:
```diff
@@ -96,6 +96,25 @@ kube-demo-undeploy: ## Uninstall the Kubernetes resources installed from the `ma
+.PHONY: install-ratelimit
+install-ratelimit:
+	@$(LOG_TARGET)
+	kubectl apply -f examples/redis/redis.yaml
```
only nit here would be to split up `redis` install with patching/updating the `EnvoyGateway` Configmap, but it can be improved later
